{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a619e187-439c-42fe-9c83-2a575088ddde",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests, json\n",
    "import time\n",
    "\n",
    "from threading import Thread\n",
    "from dateutil import parser\n",
    "from django.core.serializers.json import DjangoJSONEncoder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ad64890b",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "NAME = ['kernel', 'ozlabs', 'ffmpeg']\n",
    "INVALID_PAGE = {\n",
    "    \"detail\": \"Invalid page.\"\n",
    "}\n",
    "MAX_THREAD = 6\n",
    "PAGE_START = 104\n",
    "PAGE_NUM = 250\n",
    "TARGET_API_URL = 'http://127.0.0.1:8080/api/msr/%s'\n",
    "\n",
    "\n",
    "\n",
    "TEST_ERROR = ['''{\"original_id\": [\"people with this original id already exists.\"]}''',\n",
    "             '''{\"original_id\": [\"project with this original id already exists.\"]}''',\n",
    "             '''{\"original_id\": [\"series with this original id already exists.\"]}''',\n",
    "             '''{\"original_id\": [\"patch with this original id already exists.\"]}''',\n",
    "             '''{\"original_id\": [\"comment with this original id already exists.\"]}''']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "65f9df41-771d-402d-910b-1f62865222b4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# string = 'https://patchwork.ffmpeg.org/api/projects/'\n",
    "\n",
    "# try:\n",
    "#     found = re.search('https://patchwork\\.(.*?)\\.org', string).group(1)\n",
    "#     print(found)\n",
    "# except AttributeError:\n",
    "#     print('error')\n",
    "\n",
    "# 'A'.lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "27d6791f-3dd5-4c60-8259-b8820e726133",
   "metadata": {},
   "outputs": [],
   "source": [
    "def retrieve_basic_info(json):\n",
    "    \n",
    "    info = ['id', 'url', 'name', 'web_url']\n",
    "    result = list()\n",
    "    \n",
    "    for i in range(len(info)):\n",
    "        try:\n",
    "            result.append(json[info[i]])\n",
    "        except KeyError:\n",
    "            if info[i] == 'name':\n",
    "                result.append(json['username'])\n",
    "            elif info[i] == 'web_url':\n",
    "                result.append(json['email'])\n",
    "            else:\n",
    "                result.append(None)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "29e29063",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "def post_data(entity_type, json_data):\n",
    "    url = TARGET_API_URL %entity_type\n",
    "    payload = json.dumps(json_data, cls=DjangoJSONEncoder)\n",
    "    headers = {'Content-Type': 'application/json'}\n",
    "    response = requests.request(\"POST\", url, headers=headers, data=payload)\n",
    "    \n",
    "    if response.status_code != 201 and response.text not in TEST_ERROR:\n",
    "        print(entity_type, response.text)\n",
    "        # print(json_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c3fcaf6a-da7e-48a7-8c02-17032d311fff",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def retrieve_people_data(endpoint_name, json_people):\n",
    "    api_id, api_url, username, email = retrieve_basic_info(json_people)\n",
    "    original_id = '-'.join([endpoint_name, 'people', str(api_id)])\n",
    "    \n",
    "    item_people = {\n",
    "        'original_id': original_id,\n",
    "        'email': email,\n",
    "        'username': username,\n",
    "        'api_url': api_url\n",
    "    }\n",
    "    \n",
    "    post_data(entity_type='people', json_data=item_people)\n",
    "    \n",
    "    return original_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d9220955",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def retrieve_project_data(endpoint_name, json_project):\n",
    "    # print(f\"retrieving project: {json_project['url']}\")\n",
    "    # project info\n",
    "    project_api_id, project_api_url, project_name, project_web_url = retrieve_basic_info(json_project)\n",
    "    \n",
    "    project_repo_url = json_project['webscm_url']\n",
    "    project_list_id = json_project['list_id']\n",
    "    project_list_address = json_project['list_email']\n",
    "    project_original_id = '-'.join([endpoint_name, 'project', str(project_api_id)])\n",
    "\n",
    "    # maintainer_info\n",
    "    maintainers = json_project['maintainers']\n",
    "    maintainer_list = list()\n",
    "    for maintainer in maintainers:\n",
    "        maintainer_original_id = retrieve_people_data(endpoint_name, maintainer)\n",
    "        maintainer_list.append(maintainer_original_id)\n",
    "\n",
    "    #post project\n",
    "    item_project = {\n",
    "        'original_id': project_original_id,\n",
    "        'name': project_name,\n",
    "        'repo_url': project_repo_url,\n",
    "        'api_url': project_api_url,\n",
    "        'web_url': project_web_url,\n",
    "        'list_id': project_list_id,\n",
    "        'list_address': project_list_address,\n",
    "        'maintainers': maintainer_list\n",
    "    }\n",
    "    \n",
    "    post_data(entity_type='projects', json_data=item_project)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3c418792",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def retrieve_series_data(endpoint_name, json_series):\n",
    "    # print(f\"retrieving series: {json_series['url']}\")\n",
    "    # series info\n",
    "    series_api_id, series_api_url, series_name, series_web_url = retrieve_basic_info(json_series)\n",
    "    \n",
    "    series_created_date = parser.parse(json_series['date'])\n",
    "    series_version = json_series['version']\n",
    "    series_total = json_series['total']\n",
    "    series_received_total = json_series['received_total']\n",
    "    \n",
    "    series_original_id = '-'.join([endpoint_name, 'series', str(series_api_id)])\n",
    "    \n",
    "    series_project_api_id, _, series_project_name, _ = retrieve_basic_info(json_series['project'])\n",
    "    project_original_id = '-'.join([endpoint_name, 'project', str(series_project_api_id)])\n",
    "\n",
    "    #get cover letter content\n",
    "    if json_series['cover_letter']:\n",
    "        cover_letter_url = json_series['cover_letter']['url']\n",
    "        cover_detail = requests.get(cover_letter_url).json()\n",
    "        # series_cover_letter_content = deactivate_quote(cover_detail['content'])\n",
    "        series_cover_letter_content = cover_detail['content']\n",
    "    else:\n",
    "        series_cover_letter_content = None\n",
    "    \n",
    "    # submitter info\n",
    "    series_submitter_original_id = retrieve_people_data(endpoint_name, json_series['submitter'])\n",
    "    \n",
    "    #post series\n",
    "    item_series = {\n",
    "        'original_id': series_original_id,\n",
    "        'name': series_name,\n",
    "        'created_date': series_created_date,\n",
    "        'version': series_version,\n",
    "        'total': series_total,\n",
    "        'received_total': series_received_total,\n",
    "        'cover_letter_content': series_cover_letter_content,\n",
    "        'web_url': series_web_url,\n",
    "        'api_url': series_api_url,\n",
    "        'project_original_id': project_original_id,\n",
    "        'submitter_original_id': series_submitter_original_id,\n",
    "    }\n",
    "    \n",
    "    post_data(entity_type='series', json_data=item_series)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "051fc4b6",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def retrieve_comment_data(endpoint_name, json_comment, project_original_id, patch_original_id):\n",
    "    # print(f\"retrieving comment, id: {json_comment['web_url']}\")\n",
    "    #comment info\n",
    "    comment_api_id = json_comment['id']\n",
    "    comment_web_url = json_comment['web_url']\n",
    "    comment_msg_id = json_comment['msgid']\n",
    "    comment_msg_content = json_comment['content']\n",
    "    comment_date = parser.parse(json_comment['date'])\n",
    "    comment_subject = json_comment['subject']\n",
    "    \n",
    "    comment_original_id = '-'.join([endpoint_name, 'comment', str(comment_api_id)])\n",
    "    \n",
    "    comment_reply_to_msg_id = None\n",
    "    if 'In-Reply-To' in json_comment['headers'].keys():\n",
    "        in_reply_to = json_comment['headers']['In-Reply-To']\n",
    "        if in_reply_to[:2] == '\\n ':\n",
    "            comment_reply_to_msg_id = in_reply_to[2:]\n",
    "\n",
    "    # submitter info\n",
    "    comment_submitter_original_id = retrieve_people_data(endpoint_name, json_comment['submitter'])\n",
    "    \n",
    "    # post comment\n",
    "    item_comment = {\n",
    "        'original_id': comment_original_id,\n",
    "        'msg_id': comment_msg_id,\n",
    "        'msg_content': comment_msg_content,\n",
    "        'date': comment_date,\n",
    "        'subject': comment_subject,\n",
    "        'reply_to_msg_id': comment_reply_to_msg_id,\n",
    "        'web_url': comment_web_url,\n",
    "        'change_id': None,\n",
    "        'mailing_list_id': None,\n",
    "        'submitter_original_id': comment_submitter_original_id,\n",
    "        'patch_original_id': patch_original_id,\n",
    "        'project_original_id': project_original_id,\n",
    "    }\n",
    "    \n",
    "    post_data(entity_type='comments', json_data=item_comment)\n",
    "\n",
    "    # TODO get change id\n",
    "    # TODO get mailing list id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dd6f9f53",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def retrieve_patch_data(endpoint_name, json_patch):\n",
    "    # print(f\"retrieving patch {json_patch['url']}\")\n",
    "    #patch info\n",
    "    patch_api_id, patch_api_url, patch_name, patch_web_url = retrieve_basic_info(json_patch)\n",
    "    \n",
    "    patch_state = json_patch['state']\n",
    "    patch_date = parser.parse(json_patch['date'])\n",
    "    patch_msg_id = json_patch['msgid']\n",
    "    patch_msg_content = json_patch['content']\n",
    "    patch_code_diff = json_patch['diff']\n",
    "    \n",
    "    patch_original_id = '-'.join([endpoint_name, 'patch', str(patch_api_id)])\n",
    "    patch_project_original_id = '-'.join([endpoint_name, 'project', str(json_patch['project']['id'])])\n",
    "\n",
    "    # TODO get change id\n",
    "    # TODO get mailing list id\n",
    "\n",
    "    # get series original id\n",
    "    if json_patch['series']:\n",
    "        patch_series_api_id = json_patch['series']['id']\n",
    "        patch_series_original_id = '-'.join([endpoint_name, 'series', str(patch_series_api_id)])\n",
    "    else:\n",
    "        patch_series_original_id = None\n",
    "\n",
    "    # submitter info\n",
    "    patch_submitter_original_id = retrieve_people_data(endpoint_name, json_patch['submitter'])\n",
    "    \n",
    "    item_patch = {\n",
    "        'original_id': patch_original_id,\n",
    "        'name': patch_name,\n",
    "        'state': patch_state,\n",
    "        'date': patch_date,\n",
    "        'msg_id': patch_msg_id,\n",
    "        'msg_content': patch_msg_content,\n",
    "        'code_diff': patch_code_diff,\n",
    "        'api_url': patch_api_url,\n",
    "        'web_url': patch_web_url,\n",
    "        'change_id': None,\n",
    "        'mailing_list_id': None,\n",
    "        'series_original_id': patch_series_original_id,\n",
    "        'submitter_original_id': patch_submitter_original_id,\n",
    "        'project_original_id': patch_project_original_id,\n",
    "    }\n",
    "    \n",
    "    post_data(entity_type='patches', json_data=item_patch)\n",
    "\n",
    "    comment_url = json_patch['comments']\n",
    "    comment_list = requests.get(comment_url).json()\n",
    "    if comment_list:\n",
    "        for c in comment_list:\n",
    "            retrieve_comment_data(endpoint_name, c, patch_project_original_id, patch_original_id)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "41de5ea4-053f-4365-8714-c8d7ac341148",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main_func(api_url_base, endpoint_name, entity_type, thread_no):\n",
    "    page_num = PAGE_START + thread_no\n",
    "    response = requests.get(api_url_base %(endpoint_name, entity_type, page_num)).json()\n",
    "\n",
    "    retrieval_func = {\n",
    "        'projects': retrieve_project_data,\n",
    "        'series': retrieve_series_data,\n",
    "        'patches': retrieve_patch_data\n",
    "    }\n",
    "\n",
    "    while response != INVALID_PAGE and page_num <= PAGE_START + PAGE_NUM:\n",
    "        # print('%s: page%d started' %entity_type)\n",
    "        p_start_time = time.time()\n",
    "\n",
    "        for entity in response:\n",
    "            entity_api_url = entity['url']\n",
    "            entity_detail = requests.get(entity_api_url).json()\n",
    "\n",
    "            retrieval_func[entity_type](endpoint_name, entity_detail)\n",
    "        \n",
    "        total_time = time.time() - p_start_time\n",
    "        print('%s:\\tpage%d\\tcompleted in %.2f s' %(entity_type, page_num, total_time))\n",
    "\n",
    "        page_num += MAX_THREAD\n",
    "        response = requests.get(api_url_base %(endpoint_name, entity_type, page_num)).json()\n",
    "\n",
    "def crawl_entity(api_url_base, endpoint_name, entity_type):\n",
    "    \n",
    "    threads = [Thread(target=main_func, args=(api_url_base, endpoint_name, entity_type, thread_no)) for thread_no in range(1, MAX_THREAD + 1)]\n",
    "    for thread in threads:\n",
    "        thread.start()\n",
    "    \n",
    "    for thread in threads:\n",
    "        thread.join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7c21dcf7",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "def crawl_data(endpoint_name):\n",
    "    start_time = time.time()\n",
    "    api_url_base = 'https://patchwork.%s.org/api/%s/?page=%d'\n",
    "    # entity_types = ['projects', 'series', 'patches']\n",
    "    entity_types = ['patches']\n",
    "    \n",
    "    [crawl_entity(api_url_base, endpoint_name, entity_type) for entity_type in entity_types]\n",
    "    \n",
    "    duration = (time.time() - start_time) / 60\n",
    "    print('Retrieval completed in %.2f min' %duration)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "44a8af3a",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "patches:\tpage107\tcompleted in 245.24 s\n",
      "patches:\tpage106\tcompleted in 249.29 s\n",
      "patches:\tpage105\tcompleted in 244.69 s\n",
      "patches:\tpage110\tcompleted in 274.20 s\n",
      "patches:\tpage108\tcompleted in 277.36 s\n",
      "patches:\tpage109\tcompleted in 289.90 s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Exception in thread Thread-7:\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connection.py\", line 174, in _new_conn\n",
      "Exception in thread Exception in thread Thread-6:\n",
      "Traceback (most recent call last):\n",
      "Thread-5  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "Thread-9:\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      ":\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    conn = connection.create_connection(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\connection.py\", line 95, in create_connection\n",
      "        httplib_response = self._make_request(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 386, in _make_request\n",
      "raise err\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\connection.py\", line 85, in create_connection\n",
      "    httplib_response = self._make_request(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    six.raise_from(e, None)        \n",
      "sock.connect(sa)\n",
      "ConnectionAbortedError: [WinError 10053] 你的主机中的软件中止了一个已建立的连接。\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "    httplib_response = self._make_request(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 386, in _make_request\n",
      "self._validate_conn(conn)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 1042, in _validate_conn\n",
      "    self._validate_conn(conn)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 1042, in _validate_conn\n",
      "    httplib_response = self._make_request(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 386, in _make_request\n",
      "      File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    conn.connect()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connection.py\", line 414, in connect\n",
      "self._validate_conn(conn)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 1042, in _validate_conn\n",
      "    conn.connect()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connection.py\", line 414, in connect\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\http\\client.py\", line 1348, in getresponse\n",
      "        conn.connect()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connection.py\", line 358, in connect\n",
      "    self.sock = ssl_wrap_socket(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\ssl_.py\", line 449, in ssl_wrap_socket\n",
      "self.sock = ssl_wrap_socket(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\ssl_.py\", line 449, in ssl_wrap_socket\n",
      "    self.sock = conn = self._new_conn()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connection.py\", line 186, in _new_conn\n",
      "        ssl_sock = _ssl_wrap_socket_impl(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\ssl_.py\", line 493, in _ssl_wrap_socket_impl\n",
      "    raise NewConnectionError(\n",
      "urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPSConnection object at 0x000001601CAF4DF0>: Failed to establish a new connection: [WinError 10053] 你的主机中的软件中止了一个已建立的连接。\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\adapters.py\", line 489, in send\n",
      "ssl_sock = _ssl_wrap_socket_impl(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\ssl_.py\", line 493, in _ssl_wrap_socket_impl\n",
      "    response.begin()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\http\\client.py\", line 316, in begin\n",
      "    resp = conn.urlopen(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "    version, status, reason = self._read_status()\n",
      "    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 500, in wrap_socket\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\http\\client.py\", line 277, in _read_status\n",
      "        return ssl_context.wrap_socket(sock, server_hostname=server_hostname)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 500, in wrap_socket\n",
      "retries = retries.increment(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\retry.py\", line 592, in increment\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\socket.py\", line 669, in readinto\n",
      "    return self.sslsocket_class._create(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 1040, in _create\n",
      "    return self.sslsocket_class._create(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 1040, in _create\n",
      "    raise MaxRetryError(_pool, url, error or ResponseError(cause))\n",
      "urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='patchwork.ffmpeg.org', port=443): Max retries exceeded with url: /api/patches/3368/ (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001601CAF4DF0>: Failed to establish a new connection: [WinError 10053] 你的主机中的软件中止了一个已建立的连接。'))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\threading.py\", line 932, in _bootstrap_inner\n",
      "    return self._sock.recv_into(b)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 1241, in recv_into\n",
      "    self.do_handshake()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 1309, in do_handshake\n",
      "    self.run()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\threading.py\", line 870, in run\n",
      "        self.do_handshake()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 1309, in do_handshake\n",
      "    self._sslobj.do_handshake()\n",
      "ConnectionAbortedError: [WinError 10053] 你的主机中的软件中止了一个已建立的连接。\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\adapters.py\", line 489, in send\n",
      "return self.read(nbytes, buffer)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 1099, in read\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Timothy\\AppData\\Local\\Temp\\ipykernel_22500\\80779119.py\", line 17, in main_func\n",
      "    resp = conn.urlopen(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "        self._sslobj.do_handshake()\n",
      "ConnectionAbortedError: [WinError 10053] 你的主机中的软件中止了一个已建立的连接。\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\adapters.py\", line 489, in send\n",
      "return self._sslobj.read(len, buffer)\n",
      "ConnectionAbortedError: [WinError 10053] 你的主机中的软件中止了一个已建立的连接。\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\adapters.py\", line 489, in send\n",
      "    retries = retries.increment(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\retry.py\", line 550, in increment\n",
      "    resp = conn.urlopen(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "    resp = conn.urlopen(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\api.py\", line 73, in get\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\packages\\six.py\", line 769, in reraise\n",
      "    return request(\"get\", url, params=params, **kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\api.py\", line 59, in request\n",
      "    retries = retries.increment(    \n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\retry.py\", line 550, in increment\n",
      "return session.request(method=method, url=url, **kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\sessions.py\", line 587, in request\n",
      "    retries = retries.increment(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\retry.py\", line 550, in increment\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\packages\\six.py\", line 769, in reraise\n",
      "    httplib_response = self._make_request(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 386, in _make_request\n",
      "        resp = self.send(prep, **send_kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\sessions.py\", line 701, in send\n",
      "raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\packages\\six.py\", line 769, in reraise\n",
      "    raise value.with_traceback(tb)    r = adapter.send(request, **kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\adapters.py\", line 565, in send\n",
      "    \n",
      "raise value.with_traceback(tb)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "        self._validate_conn(conn)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 1042, in _validate_conn\n",
      "httplib_response = self._make_request(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    raise ConnectionError(e, request=request)\n",
      "requests.exceptions.ConnectionError: HTTPSConnectionPool(host='patchwork.ffmpeg.org', port=443): Max retries exceeded with url: /api/patches/3368/ (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001601CAF4DF0>: Failed to establish a new connection: [WinError 10053] 你的主机中的软件中止了一个已建立的连接。'))\n",
      "    httplib_response = self._make_request(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 386, in _make_request\n",
      "        six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "conn.connect()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connection.py\", line 414, in connect\n",
      "        self._validate_conn(conn)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 1042, in _validate_conn\n",
      "self.sock = ssl_wrap_socket(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\ssl_.py\", line 449, in ssl_wrap_socket\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    conn.connect()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\connection.py\", line 414, in connect\n",
      "        ssl_sock = _ssl_wrap_socket_impl(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\ssl_.py\", line 493, in _ssl_wrap_socket_impl\n",
      "httplib_response = conn.getresponse()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\http\\client.py\", line 1348, in getresponse\n",
      "    self.sock = ssl_wrap_socket(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\ssl_.py\", line 449, in ssl_wrap_socket\n",
      "    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)    \n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 500, in wrap_socket\n",
      "response.begin()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\http\\client.py\", line 316, in begin\n",
      "    ssl_sock = _ssl_wrap_socket_impl(        \n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\urllib3\\util\\ssl_.py\", line 493, in _ssl_wrap_socket_impl\n",
      "version, status, reason = self._read_status()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\http\\client.py\", line 277, in _read_status\n",
      "return self.sslsocket_class._create(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 1040, in _create\n",
      "        line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\socket.py\", line 669, in readinto\n",
      "self.do_handshake()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 1309, in do_handshake\n",
      "    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 500, in wrap_socket\n",
      "    return self._sock.recv_into(b)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 1241, in recv_into\n",
      "        self._sslobj.do_handshake()\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionAbortedError(10053, '你的主机中的软件中止了一个已建立的连接。', None, 10053, None))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\threading.py\", line 932, in _bootstrap_inner\n",
      "return self.sslsocket_class._create(\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 1040, in _create\n",
      "    self.run()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\threading.py\", line 870, in run\n",
      "    return self.read(nbytes, buffer)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 1099, in read\n",
      "    self.do_handshake()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\ssl.py\", line 1309, in do_handshake\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Timothy\\AppData\\Local\\Temp\\ipykernel_22500\\80779119.py\", line 17, in main_func\n",
      "    return self._sslobj.read(len, buffer)\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionAbortedError(10053, '你的主机中的软件中止了一个已建立的连接。', None, 10053, None))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\threading.py\", line 932, in _bootstrap_inner\n",
      "    self._sslobj.do_handshake()\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionAbortedError(10053, '你的主机中的软件中止了一个已建立的连接。', None, 10053, None))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\threading.py\", line 932, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\threading.py\", line 870, in run\n",
      "    self.run()\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\threading.py\", line 870, in run\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\api.py\", line 73, in get\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Timothy\\AppData\\Local\\Temp\\ipykernel_22500\\80779119.py\", line 17, in main_func\n",
      "        return request(\"get\", url, params=params, **kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\api.py\", line 59, in request\n",
      "self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Timothy\\AppData\\Local\\Temp\\ipykernel_22500\\80779119.py\", line 19, in main_func\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\api.py\", line 73, in get\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\sessions.py\", line 587, in request\n",
      "    return request(\"get\", url, params=params, **kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\api.py\", line 59, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\sessions.py\", line 701, in send\n",
      "  File \"C:\\Users\\Timothy\\AppData\\Local\\Temp\\ipykernel_22500\\3480833491.py\", line 48, in retrieve_patch_data\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\sessions.py\", line 587, in request\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\adapters.py\", line 547, in send\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\sessions.py\", line 701, in send\n",
      "      File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\api.py\", line 73, in get\n",
      "raise ConnectionError(err, request=request)\n",
      "requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionAbortedError(10053, '你的主机中的软件中止了一个已建立的连接。', None, 10053, None))\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\adapters.py\", line 547, in send\n",
      "    return request(\"get\", url, params=params, **kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\api.py\", line 59, in request\n",
      "        raise ConnectionError(err, request=request)\n",
      "requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionAbortedError(10053, '你的主机中的软件中止了一个已建立的连接。', None, 10053, None))\n",
      "return session.request(method=method, url=url, **kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\sessions.py\", line 587, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\sessions.py\", line 701, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"C:\\Users\\Timothy\\anaconda3\\envs\\MSR_MDB\\lib\\site-packages\\requests\\adapters.py\", line 547, in send\n",
      "    raise ConnectionError(err, request=request)\n",
      "requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionAbortedError(10053, '你的主机中的软件中止了一个已建立的连接。', None, 10053, None))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "patches:\tpage114\tcompleted in 172.16 s\n",
      "patches:\tpage116\tcompleted in 174.61 s\n",
      "patches:\tpage122\tcompleted in 194.24 s\n",
      "patches:\tpage120\tcompleted in 223.60 s\n",
      "patches:\tpage128\tcompleted in 170.59 s\n",
      "patches:\tpage126\tcompleted in 181.06 s\n",
      "patches:\tpage134\tcompleted in 135.49 s\n",
      "patches:\tpage132\tcompleted in 146.78 s\n",
      "patches:\tpage138\tcompleted in 164.69 s\n",
      "patches:\tpage140\tcompleted in 335.72 s\n",
      "patches:\tpage144\tcompleted in 189.93 s\n",
      "patches:\tpage146\tcompleted in 177.34 s\n",
      "patches:\tpage150\tcompleted in 172.80 s\n",
      "patches:\tpage152\tcompleted in 178.00 s\n",
      "patches:\tpage156\tcompleted in 194.46 s\n",
      "patches:\tpage158\tcompleted in 175.86 s\n",
      "patches:\tpage162\tcompleted in 265.05 s\n",
      "patches:\tpage164\tcompleted in 169.83 s\n",
      "patches:\tpage170\tcompleted in 168.84 s\n",
      "patches:\tpage168\tcompleted in 182.95 s\n",
      "patches:\tpage176\tcompleted in 189.87 s\n",
      "patches:\tpage174\tcompleted in 196.57 s\n",
      "patches:\tpage180\tcompleted in 222.12 s\n",
      "patches:\tpage182\tcompleted in 241.99 s\n",
      "patches:\tpage186\tcompleted in 169.48 s\n",
      "patches:\tpage188\tcompleted in 220.02 s\n",
      "patches:\tpage192\tcompleted in 194.97 s\n",
      "patches:\tpage194\tcompleted in 185.91 s\n",
      "patches:\tpage200\tcompleted in 130.67 s\n",
      "patches:\tpage198\tcompleted in 184.34 s\n",
      "patches:\tpage206\tcompleted in 214.85 s\n",
      "patches:\tpage204\tcompleted in 245.67 s\n",
      "patches:\tpage212\tcompleted in 180.62 s\n",
      "patches:\tpage210\tcompleted in 214.24 s\n",
      "patches:\tpage218\tcompleted in 212.89 s\n",
      "patches:\tpage216\tcompleted in 230.61 s\n",
      "patches:\tpage224\tcompleted in 140.12 s\n",
      "patches:\tpage222\tcompleted in 158.29 s\n",
      "patches:\tpage230\tcompleted in 158.01 s\n",
      "patches:\tpage228\tcompleted in 165.24 s\n",
      "patches:\tpage236\tcompleted in 207.57 s\n",
      "patches:\tpage234\tcompleted in 267.93 s\n",
      "patches:\tpage242\tcompleted in 207.22 s\n",
      "patches:\tpage240\tcompleted in 167.80 s\n",
      "patches:\tpage248\tcompleted in 215.33 s\n",
      "patches:\tpage246\tcompleted in 205.03 s\n",
      "patches:\tpage254\tcompleted in 251.68 s\n",
      "patches:\tpage252\tcompleted in 207.25 s\n",
      "patches:\tpage260\tcompleted in 208.66 s\n",
      "patches:\tpage258\tcompleted in 186.51 s\n",
      "patches:\tpage266\tcompleted in 183.85 s\n",
      "patches:\tpage264\tcompleted in 185.49 s\n",
      "patches:\tpage272\tcompleted in 197.80 s\n",
      "patches:\tpage270\tcompleted in 184.61 s\n",
      "patches:\tpage278\tcompleted in 167.54 s\n",
      "patches:\tpage276\tcompleted in 165.52 s\n",
      "patches:\tpage284\tcompleted in 229.78 s\n",
      "patches:\tpage282\tcompleted in 216.18 s\n",
      "patches:\tpage288\tcompleted in 162.04 s\n",
      "patches:\tpage290\tcompleted in 212.65 s\n",
      "patches:\tpage294\tcompleted in 188.11 s\n",
      "patches:\tpage296\tcompleted in 213.11 s\n",
      "patches:\tpage300\tcompleted in 225.53 s\n",
      "patches:\tpage302\tcompleted in 187.46 s\n",
      "patches:\tpage308\tcompleted in 174.25 s\n",
      "patches:\tpage306\tcompleted in 208.26 s\n",
      "patches:\tpage314\tcompleted in 182.26 s\n",
      "patches:\tpage312\tcompleted in 169.47 s\n",
      "patches:\tpage320\tcompleted in 163.10 s\n",
      "patches:\tpage318\tcompleted in 163.89 s\n",
      "patches:\tpage326\tcompleted in 165.82 s\n",
      "patches:\tpage324\tcompleted in 184.50 s\n",
      "patches:\tpage332\tcompleted in 207.31 s\n",
      "patches:\tpage330\tcompleted in 209.22 s\n",
      "patches:\tpage336\tcompleted in 158.43 s\n",
      "patches:\tpage338\tcompleted in 203.43 s\n",
      "patches:\tpage342\tcompleted in 189.42 s\n",
      "patches:\tpage344\tcompleted in 187.58 s\n",
      "patches:\tpage348\tcompleted in 150.45 s\n",
      "patches:\tpage350\tcompleted in 172.59 s\n",
      "patches:\tpage354\tcompleted in 148.33 s\n",
      "Retrieval completed in 136.69 min\n"
     ]
    }
   ],
   "source": [
    "crawl_data(NAME[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3522c36-a91e-409b-8d69-8d79972f2c35",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
